### PR TITLE
JAMES-2557 Finally deprecate MailAddress::nullSender

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: core/src/main/java/org/apache/james/core/MaybeSender.java [hunk 1]
```diff
@@ -24,10 +24,33 @@
...
+import org.slf4j.LoggerFactory;
+
 import com.google.common.base.MoreObjects;
 import com.google.common.collect.ImmutableList;
 
 public class MaybeSender {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MaybeSender.class);
+
+    public static MaybeSender getMailSender(String sender) {
+        if (sender == null || sender.trim().length() <= 0) {
```
Comment (mbaechler) [inline core/src/main/java/org/apache/james/core/MaybeSender.java]: I don't think a length can be `<0`

File: core/src/main/java/org/apache/james/core/MaybeSender.java [hunk 2]
```diff
@@ -24,10 +24,33 @@
...
+            return MaybeSender.nullSender();
+        }
+        if (sender.equals(MailAddress.NULL_SENDER_AS_STRING)) {
+            return MaybeSender.nullSender();
+        }
+        try {
+            return MaybeSender.of(new MailAddress(sender));
+        } catch (AddressException e) {
+            // Should never happen as long as the user does not modify the header by himself
+            LOGGER.error("Unable to parse the sender address {}, so we fallback to a null sender", sender, e);
```
Comment (mbaechler) [inline core/src/main/java/org/apache/james/core/MaybeSender.java]: error looks to much to me, I propose warn

File: core/src/main/java/org/apache/james/core/MaybeSender.java (hunk 1 above)
Comment (chibenwa) [reply core/src/main/java/org/apache/james/core/MaybeSender.java]: replaced by a `isEmpty` call

**Comment by chibenwa** [comment]
```
[02f2783c7dec58360440824877f26be90de04225] [ERROR] queueShouldHandleSender  Time elapsed: 0.012 s  <<< ERROR!
[02f2783c7dec58360440824877f26be90de04225] java.lang.NullPointerException
[02f2783c7dec58360440824877f26be90de04225] 	at org.apache.james.queue.api.MailQueueContract.enQueue(MailQueueContract.java:64)
[02f2783c7dec58360440824877f26be90de04225] 	at org.apache.james.queue.api.MailQueueContract.queueShouldHandleSender(MailQueueContract.java:97)
[02f2783c7dec58360440824877f26be90de04225] 
[02f2783c7dec58360440824877f26be90de04225] [ERROR] queueShouldHandleNoSender  Time elapsed: 0.006 s  <<< ERROR!
[02f2783c7dec58360440824877f26be90de04225] java.lang.NullPointerException
[02f2783c7dec58360440824877f26be90de04225] 	at org.apache.james.queue.api.MailQueueContract.enQueue(MailQueueContract.java:64)
[02f2783c7dec58360440824877f26be90de04225] 	at org.apache.james.queue.api.MailQueueContract.queueShouldHandleNoSender(MailQueueContract.java:112)
[02f2783c7dec58360440824877f26be90de04225] 
[02f2783c7dec58360440824877f26be90de04225] [INFO] 
[02f2783c7dec58360440824877f26be90de04225] [INFO] Results:
[02f2783c7dec58360440824877f26be90de04225] [INFO] 
[02f2783c7dec58360440824877f26be90de04225] [ERROR] Errors: 
[02f2783c7dec58360440824877f26be90de04225] [ERROR]   MailQueueContract.queueShouldHandleNoSender:112->enQueue:64 » NullPointer
[02f2783c7dec58360440824877f26be90de04225] [ERROR]   MailQueueContract.queueShouldHandleSender:97->enQueue:64 » NullPointer
```

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
`SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:242 » ConditionTimeout`

test this please

**Comment by chibenwa** [comment]
test this please

File: core/src/test/java/org/apache/james/core/MaybeSenderTest.java [hunk 3]
```diff
@@ -155,4 +156,29 @@ void asStringWithDefaultShouldReturnDefaultWhenNullSender() {
             .isEqualTo("default");
     }
 
+
+
+    @Test
+    void getMailSenderShouldReturnNullSenderWhenNullSender() {
```
Comment (trantienduchn) [inline core/src/test/java/org/apache/james/core/MaybeSenderTest.java]: missed when sender is `null` or `empty string`, `bad formated sender value` tests. could use parameterized test

File: protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/AbstractSenderAuthIdentifyVerificationRcptHook.java [hunk 4]
```diff
@@ -45,28 +47,36 @@
...
-        return null;
+    private boolean isAnonymous(MaybeSender maybeSender) {
+        return maybeSender == null || maybeSender.isNullSender();
+    }
+
+    private boolean hasRightUsername(MaybeSender maybeSender, SMTPSession session) {
+        Preconditions.checkArgument(!maybeSender.isNullSender());
+
+        String authUser = session.getUser().toLowerCase(Locale.US);
+        String username = getUser(maybeSender.get());
```
Comment (trantienduchn) [inline protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/AbstractSenderAuthIdentifyVerificationRcptHook.java]: do we need to convert string `toLowerCase()` here?

File: protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/DataCmdHandler.java [hunk 5]
```diff
@@ -1,234 +1,234 @@
...
+import org.apache.james.protocols.smtp.dsn.DSNStatus;
+import org.apache.james.util.MDCBuilder;
+
+import com.google.common.collect.ImmutableSet;
+
+
+/**
+ * handles DATA command
+ */
+public class DataCmdHandler implements CommandHandler<SMTPSession>, ExtensibleHandler {
```
Comment (trantienduchn) [inline protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/DataCmdHandler.java]: can you rework this file? I see whole file change, but we just need only one instead right?

File: protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/AbstractSenderAuthIdentifyVerificationRcptHook.java (hunk 4 above)
Comment (chibenwa) [reply protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/AbstractSenderAuthIdentifyVerificationRcptHook.java]: Well, most of this logic is not covered by tests, changing it too much do not seems like a good idea.

File: protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/DataCmdHandler.java (hunk 5 above)
Comment (chibenwa) [reply protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/DataCmdHandler.java]: > can you rework this file? 

No, it is not the intent of this commit, whose title is `JAMES-2557 DataCmdHandler line breaks`

I did **no code modification**.

>  I see whole file change, but we just need only one instead right?

See the commit view detailed here https://github.com/linagora/james-project/pull/2003/commits

File: protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/AbstractSenderAuthIdentifyVerificationRcptHook.java (hunk 4 above)
Comment (chibenwa) [reply protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/AbstractSenderAuthIdentifyVerificationRcptHook.java]: https://ci.linagora.com/linagora/lgs/openpaas/james/issues/961

File: core/src/test/java/org/apache/james/core/MaybeSenderTest.java (hunk 3 above)
Comment (chibenwa) [reply core/src/test/java/org/apache/james/core/MaybeSenderTest.java]: Indeed we can add the missing edge cases.

Concerning parametrized tests, it gets the syntax harder for little benefits. I'd rather not do it.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**